### ISSUE TITLE
BUG: Endogenous creation of `bg_id`

### ISSUE THREAD

**Issue by MImmesberger** [iza-institute-of-labor-economics/gettsim#763]
The current creation of `bg_id` is wrong because children with `eigenbedarf_gedeckt=True` will always have a different `bg_id` than their parents, even if the parents have `eigenbedarf_gedeckt=True` as well.

### Proposed solution

Parents should have the same `bg_id` as their children if all of them have `eigenbedarf_gedeckt=True`.

**Comment by hmgaudecker** [comment]
Sounds okay as a short-term hack. But we'll have to be careful that
1. both parents always have the same value of `eigenbedarf_gedeckt` 
2. we treat children with `eigenbedarf_gedeckt == False` correctly (thinking of patchwork here).

Ultimately, however, `eigenbedarf_gedeckt` really needs to go -- it is just a synonym for "separate bg's" and with the suggested change, its name is not even related closely  to the content anymore.

I am rather convinced now that we won't be able to solve all of this with one pass of GETTSIM -- unless we find a way of working with temporary/candidate variables in a subgraph.

What I mean is that we'll need to check multiple things to determine the `bg_id`, which depend on the configurations of `bg`'s within a household. So we"ll have to check various candidate configurations (children out? Bürgergeld even relevant? Possible `wthh_id`s?), which determine the payments. Example from going through budget sets of exemplary households:
 
> Mir ist folgendes Problem aufgefallen:
> 
> - an dem Punkt, an dem Haushalt von "Elternteil ALG 2 + Kind WG" wechselt zu "gesamter Haushalt WG",  fällt das Einkommen des Haushaltes ein gutes Stück (obwohl das Bruttoeinkommen steigt)
> - Der Grund ist, dass beim Elternteil ALG 2 für eine Person verglichen wird mit WG für eine Person. Hier ist dann irgendwann WG günstiger (auch schon recht früh durch den Kindergeldübertrag). Ausgezahlt wird das Wohngeld dann aber nicht für zwei Einpersonenhaushalte (wie in der Günstigerprüfung angenommen), sondern das Wohngeld wird für beide Personen berechnet, was weniger ist und das verfügbare Einkommen reduziert.

In the end, this is sort of the same as calculating candidate benefits before checking eligibility / priority. We solve that by giving things names like `xxx_vor_vorrang_check`. But that is ugly and not feasible for this much more complex case. There might be a better way of solving both of these. 

My suggestion would be to do a quick hack here as you suggest, do the transition to a hierarchical scheme, and revisit.

**Comment by MImmesberger** [comment]
For future reference, here is the logic that I'm using in the policy project:

The problem of setting the `bg_id` boils down to 1) calculation `eigenbedarf_gedeckt` and then setting `bg_id` correctly and then 2) making sure that no BG covers their own needs individually but not together (because Wohngeld for the group is lower than the sum of the individual Wohngeld claims).

I used the following procedure:

1. Set `bg_id` to be different for each child and the parents (i.e. assuming that each child covers it's own needs, e.g. by setting `eigenbedarf_gedeckt=True` for everyone).
2. Calculate whether Eigenbedarf is covered given the `bg_id` from (1) and then set
   another candidate `bg_id` based on the results.
3. Then, calculate for this `bg_id` from (2) again whether the total Eigenbedarf is covered.
4. Set the final `bg_id` based on the following logic:

Potential Outcomes:
- (A) In step (2), **parental Eigenbedarf is covered**
    - Children are part of parental BG in GETTSIM call of (3), it doesn't matter whether
  their Eigenbedarf was covered.
    - (A.1.) In step (3), the **BG doesn't cover the Eigenbedarf**
        - (A.1.1.) In step (2), **children could cover their Eigenbedarf**
            - Accept the `bg_id` set in (1).
        - (A.1.2.) In step (2), **children couldn't cover their Eigenbedarf**
            - Accept the `bg_id` that was used in (3).
    - (A.2.) In step (3), the **BG covers the Eigenbedarf**
        - Accept the `bg_id` that was used in (3).
- (B.) In step (2), **parental Eigenbedarf is not covered**
    - (B.1). In step (2), **children couldn't cover their Eigenbedarf**
        - Children are part of the parental BG in GETTSIM at call of step (3).
        - (B.1.1.) In step (3), **the BG doesn't cover their Eigenbedarf**
            - Accept `bg_id` that was used in (3).
        - (B.1.2.) In step (3), **BG does cover its Eigenbedarf**
        - This should never happen.
    - (B.2.) In step (2), **children could cover their Eigenbedarf**
        - Children are part of a different BG than parents
        - (B.2.1.) In step (3), **parents cannot cover their Eigenbedarf**
            - Accept the `bg_id` that was used in (3) and (2).
        - (B.2.2.) In step (3), **parents can cover their Eigenbedarf**
            - This should never happen.

To calculate whether Eigenbedarf is covered, I use the `wohngeld_kinderzuschl_vorrang_bg` target of GETTSIM.

**Comment by hmgaudecker** [comment]
Sounds very good! I'd like to suggest a different strategy, I won't describe the algorithm quite as elaborately as you did. But it will be better to draw that on a board, anyhow.

1. Calculate Bürgergeld claim for the whole `fg`. 
   1. If there is none, set a flag to ignore these in all the rest. Everything below can never matter then; better not to have to think about it.
   2. If there is one, continue with the logic in 2.. I.e., we focus on cases where *someone* is eligible for Bürgergeld for sure.
2. Check whether some children cover their needs and set `bg_id`s. Do *not* use `wohngeld_kinderzuschl_vorrang_bg` for this check but something like what Christian posted [here](https://github.com/iza-institute-of-labor-economics/gettsim/issues/758#issuecomment-2118846703). 

   The reason is that we do not want to have Wohngeld in that check. And Kinderzuschlag even less so (can only be paid if living with adults in the same `bg`). If children are taken out of a `bg` only because of Kinderwohngeld, doing so is optional and probably quite rare. 

   When children cover their needs, set a flag that we do not need to do any Bürgergeld priority checks for them below / they will always be in the `wthh` which does not receive Bürgergeld.

3. Calculate Wohngeld. In order to prevent the [issue quoted above](https://github.com/iza-institute-of-labor-economics/gettsim/issues/763#issuecomment-2148817621), we *might* need to do that twice:
   1. Setting the `wthh_id` given the current configuration of `bg_id`s, assuming parents receive Bürgergeld.
   2. Setting the `wthh_id` assuming that the entire `fg` does not receive Bürgergeld (most likely this can be skipped because we exclude these cases in 1.i. -- the check would be the same, right? We should think through it before leaving it out, though)

This prevents the circular logic (split up fg -- do individual checks in which nobody is eligible for Bürgergeld -- merge again -- split up fg -- ...) that plagued us in that issue quoted above because of the up-front check in 1.i

Do you see any flaw with it?

**Comment by MImmesberger** [comment]
That's great, it's much easier to think about the problem this way! In my algorithm above, I inadvertently tried to account for Kinderwohngeld which made this super complex.

I think (1.i) doesn't work. The whole `fg` may not be eligible for ALG2 because children receive high Unterhalt, but in the end, parents may be eligible for it after children were removed from the BG. But your algorithm doesn't rely on this.

The rest should work, but it hinges on the Bürgergeld Regelbedarfe to be a linear function of the children included/excluded in the BG (i.e. sum of Regelbedarf of two children BGs = sum of Regelbedarf of one BG of the same two children) which is true I think but we may have to keep that in mind for the future.

I would slightly adjust the wording of (3): After doing (2) we always end up with exactly two candidate `wthh_id` configurations: i) the group of children that covered their needs in (2) receives Wohngeld, parents and the other children don't (different `wthh_id`), ii) the group of children that covered their needs in (2) receives Wohngeld, parents and the other children do as well (same `wthh_id`).

**Comment by hmgaudecker** [comment]
Cool! 

No time right now, but: How about leaving child income out in 1. ? Likely means we'll have to do 3.ii, but should do the job?

**Comment by hmgaudecker** [comment]
I edited the above algorithm using that idea. Should be clearer than discussing back and forth here.

**Comment by hmgaudecker** [comment]
> The rest should work, but it hinges on the Bürgergeld Regelbedarfe to be a linear function of the children included/excluded in the BG (i.e. sum of Regelbedarf of two children BGs = sum of Regelbedarf of one BG of the same two children) which is true I think but we may have to keep that in mind for the future.

I don't think so. Each child is its own `bg` (different from Wohngeld!) , but that never matters because by the previous check, these kids do cover their needs.

**Comment by MImmesberger** [comment]
> Say the check in 1. shows a Bürgergeld claim, but then children are taken out because of their income and suddenly the parents don't have a claim anymore for themselves? 

I think this is what I meant when I was saying that it works because Regelbedarf is linear in BG members (just didn't realise that Bürgergeld is the relevant object here, not Regelbedarf). This situation can happen since `anz_kinder_bis_17_bg` determines the relevant SGB II income but I would argue in this situation the whole `fg` gets Wohngeld?

But of course, you're right. There are probably corner cases that we haven't thought about yet..

**Comment by hmgaudecker** [comment]
Bürgergeld == Regelbedarf + eligibility check, right? Just need to be careful whether we need the check at some point or not.

> I would argue in this situation the whole fg gets Wohngeld?

Hopefully :smile:

**Comment by MImmesberger** [comment]
Let me elaborate on step 3 a bit more (sorry that this is super long but I think it really helps to spell it out).

When leaving step 2, the `bg_id` depends on the `wthh_id` to the extent that we want children to be in the same BG as their parents if they have the same `wthh_id` (no matter the results from step 2!). If `wthh_id` is different (i.e. parents receive Bürgergeld, some children receive Wohngeld), the `bg_id` should be different (and is determined by Step 2).

So the crucial step is determining who receives Wohngeld and thereby setting `wthh_id`. We know that every child who has covered their needs individually **must** be in the Wohngeld group (Vorrangprüfung). For parents and and the other children its unclear because the Vorrangprüfung is not a perfect predictor of the Günstigerprüfung (the latter making the ultimate decision about Wohngeld vs BüG, the Günstigerprüfung ist just the Vorrangprüfung with Wohngeld and KiZ counted as potential income).

So there are two candidate constellations:

1. The parents (and the children who did not cover their needs individually) fail the Günstigerprüfung. Then, there are two `wthh_id`s for that household, and children who did cover their needs are in a different BG than parents.
2. The parents (and the children who did not cover their needs individually) pass the Günstigerprüfung. The whole `fg` will share the same `wthh_id` and `bg_id`.

As far as I can tell, these are the only two options.

The difficulty now is how to exactly model the Günstigerprüfung. I propose the following algorithm:

1. Set `wthh_id` and `bg_id` following the description of the first scenario. Run GETTSIM on the targets `wohngeld_vorrang_bg`, `wohngeld_kinderzuschl_vorrang_bg` and other targets that are needed for computing disposable income. If the parental BG did not pass the Günstigerprüfung, accept candidate solution (1). If they did pass, go to the next step.
2. Set `wthh_id` and `bg_id` following the description of the second scenario. Run GETTSIM on the same targets as in the point above. If they pass the Günstigerprüfung, accept candidate solution (2).

Now we need to handle those cases where they pass the check in the second point, but fail the check in the third point (due to nonlinearity of Wohngeld). This likely affects only a tiny range of SGB II - relevant income (this is a different jump in disposable income than the one Christian was mentioning, this one here should affect a smaller income range if I understood it correctly), but afaik we have no idea what **should** happen here. I think it would be most compliant with the law to accept candidate solution (1) here because the parents and children than form one BG in candidate solution (1) are still in the income range where they did not cover their own needs as a BG, i.e. have a rightful claim on BüG. Since candidate (2) is not a viable option they cannot receive WoG. Thus, we accept candidate solution (1) and act as if they failed the Günstigerprüfung of step 1.

**Comment by hmgaudecker** [comment]
Thanks! Sounds very exhaustive indeed. I need to work through that very slowly at some point.

Would you be able to come up with some test cases that illustrate these points? That would help clarity tremendously. Without knowing exactly what we are trying to achieve, it is just hard to follow multi-step reasoning.

If it helps, add intermediate results for test cases. If there are multiple possible outcomes for a test (you mention it is unclear in some cases what *should* happen), leave those open and it will be great to discuss them.

**Comment by MImmesberger** [comment]
Yes, will do that!

I updated my reasoning above slightly. I noticed that there might be situation where the Günstigerprüfung says WoG>BüG for parents and children individually and for parents and children in one BG, but in fact parent receiving BüG and child receiving WoG yields the higher disposable income. Hence, the Günstigerprüfung must be a bit more sophisticated (by considering the whole `fg`) than it is currently.

**Comment by MImmesberger** [comment]
@hmgaudecker I haven't started yet with the tests. But I hope you can help here in this case without having the test cases in front of you:

I just realised that the path dependence of modelling the Günstigerprüfung is what makes this so complicated. When we only model the Vorrangprüfung, the algorithm above should yield the correct result.

If we want to model the eingeschränktes Wahlrecht (=Günstigerprüfung) of Wohngeld, i.e. people can get Wohngeld if the transfer is higher than BüG, setting the `bg_id` and `wthh_id` gets complicated once we have multiple `fg`s in one household and we want to maximize household income. Then, we would need to compare multiple `wthh_id` candidates on the household level and we're back with an undetermined amount of GETTSIM calls. I think there are two solutions:

1. We do the Günstigerprüfung on the `fg` level. The downside is that this creates jumps in the disposable income of the household (because what maximizes income of the `fg` level does not necessarily maximize income at the `hh` level; think, for example, about two `fg`s and for both individually, it's better to receive Wohngeld than BüG, but they're eligible for both. Combining them may make them worse off.).
2. We do the same as with Kinderwohngeld: Only rely on the Vorrangprüfung, neglecting the Günstigerprüfung. This would make sense given that the Günstigerprüfung is probably (?) only done upon explicit request.

**Comment by MImmesberger** [comment]
![dagitty-model-2](https://github.com/iza-institute-of-labor-economics/gettsim/assets/74215010/25232fed-9fac-47da-ab7a-02e002119f3b)

(`...` are placeholders for many other functions that come in between which I won't touch) 

This is my current plan. The three key targets are `wthh_id_endogen`, `bg_id_endogen` and `wohngeld_kinderzuschl_statt_arbeitsl_geld_2_endogen`.

If the user does not set `wthh_id` and `bg_id` themselves, GETTSIM creates the IDs such that children who can cover their needs are not part of the parental BG (this is basically the status quo, just with the adjustment that if the whole `fg` covers their needs, they form a BG again).

The `wohngeld_kinderzuschl_statt_arbeitsl_geld_2` bool **must** be set by the user, even if `wthh_id` and `bg_id` are provided. There are ways to rationalise `wohngeld_kinderzuschl_statt_arbeitsl_geld_2` from the inputs if there is only on `fg` in the household (I think) but it's not possible in general. 

The incorrect Günstigerprüfung in the current version drops out without replacement, i.e. if a user wants to compute Wohngeld/ALG2 without guessing `wohngeld_kinderzuschl_statt_arbeitsl_geld_2`, the user must run GETTSIM twice.

Feel free to suggest other target names if you think something is unclear!

**Comment by hmgaudecker** [comment]
Thanks! A couple of reactions:
- We should offer a check whether `wohngeld_kinderzuschl_statt_arbeitsl_geld_2_endogen` and `wohngeld_kinderzuschl_statt_arbeitsl_geld_2` line up, same for the ids. We should then alert users that results are likely wrong
- `eigenbedarf_gedeckt` looks like an input -- that should not be necessary?
- As discussed in the other thread, I am no fan (anymore) of things like `xxx_nach_vermögensprüfung`. Anything that allows us to make absolute checks we should do upfront and then carry along an `anspruchsberechtigt` or the like.

**Comment by MImmesberger** [comment]
> `eigenbedarf_gedeckt` looks like an input -- that should not be necessary?

No, that's an output generated via 

def eigenbedarf_gedeckt(
    arbeitsl_geld_2_regelbedarf_m: float,
    _arbeitsl_geld_2_nettoeink_ohne_transfers_m: float,
    kindergeld_zur_bedarfsdeckung_m: float,
    kind_unterh_erhalt_m: float,
    unterhaltsvors_m: float,
) -> bool:
    """Check if SGB II needs are covered.
    ...

    """
    return (
        arbeitsl_geld_2_regelbedarf_m
        <= _arbeitsl_geld_2_nettoeink_ohne_transfers_m
        + kindergeld_zur_bedarfsdeckung_m
        + kind_unterh_erhalt_m
        + unterhaltsvors_m
    )

> As discussed in the other thread, I am no fan (anymore) of things like xxx_nach_vermögensprüfung. Anything that allows us to make absolute checks we should do upfront and then carry along an anspruchsberechtigt or the like.

Definitely, I should have added that here as well.

**Comment by hmgaudecker** [comment]
> > eigenbedarf_gedeckt looks like an input -- that should not be necessary?

> No, that's an output generated via

Great, just wasn't obvious from the graph. I think we should make that `arbeitsl_geld_2_eigenbedarf_gedeckt`

**Comment by hmgaudecker** [comment]
Though that will be clear very soon, anyhow. So no big deal.

**Comment by MImmesberger** [comment]
> We should offer a check whether wohngeld_kinderzuschl_statt_arbeitsl_geld_2_endogen and wohngeld_kinderzuschl_statt_arbeitsl_geld_2 line up, same for the ids. We should then alert users that results are likely wrong

Just a heads up: This would blow up the modules further. This is because the user influences the calculation of e.g. `wohngeld_anspruchshöhe_wthh` by setting `wthh_id` and `wohngeld_kinderzuschl_statt_arbeitsl_geld_2`. My current plan was to create them endogenously such that they are useful for calculating the endogenous `bg_id`s. This ensured that we only need, e.g. `wohngeld_anspruchshöhe_wthh` and `wohngeld_anspruchshöhe_fg` (and all the upstream `_fg` and `_wthh` functions).

But I agree, that this would be important and I think the additional lines are worth it.

**Comment by hmgaudecker** [comment]
> My current plan was to create them endogenously

What exactly does "them" refer to in this sentence?

**Comment by MImmesberger** [comment]
Sorry, I was referring to`wthh_id`, `bg_id`. If they are not specified by the user, I currently create them such that children who cover their needs are not in the parental BG. Then I compare this candidate to the situation where the whole FG forms the BG.

**Comment by hmgaudecker** [comment]
Sure, but we expect them to be set by the user along with `wohngeld_kinderzuschl_statt_arbeitsl_geld_2`, right? Maybe I don't quite get the whole structure yet. Maybe it is also not quite possible what we are trying to do and we'll just need to have a function calling `compute_taxes_and_transfers` multiple times to get at the correct configuration?

**Comment by MImmesberger** [comment]
Yes, my point is just that the warning you were suggesting above is not possible with the current structure of #778.

If we want such a warning, we could do this via multiple `compute_taxes_and_transfers` calls.

**Comment by hmgaudecker** [comment]
Linking a [comment on Unterhaltsvorschuss](https://github.com/iza-institute-of-labor-economics/gettsim/issues/479#issuecomment-1377742006), which we should double-check here.
